### PR TITLE
Added SOCKS4 support.

### PR DESCRIPTION
**Pull request by RoKu1**
Supports: #39 
SOCKS4 support with and without userid.
Both `Connect` and `Bind` supported.
RFC Followed: [SOCKS4 RFC](https://www.openssh.com/txt/socks4.protocol)

### CODE REVIEW

**Comment by RoKu1** [comment]
Hi @sticnarf , this is a complete socks4 support addition, with both Connect and Bind. Also, additionally UserID support is added. Code is organised as per existing flow, and separated out in a new file to avoid confusion.

File: src/lib.rs [hunk 1]
```diff
@@ -260,6 +261,7 @@ impl<'a> Authentication<'a> {
 
 mod error;
 pub mod tcp;
+pub mod tcp_socks4;
```
Comment (sticnarf) [inline src/lib.rs:264]: I prefer to put the socks4 types directly in the `tcp` mod.

To better organize the code, maybe we can move the current socks5 code to tcp/socks5 and put the sock4 code in tcp/socks4. Then, re-export the useful types to the tcp mod for convenience.

**Review by sticnarf — COMMENTED** [review]
Looks great!

I just fixed the broken Github Action. Can you please rebase so we can give it a test on CI again?

File: src/lib.rs (hunk 1 above)
Comment (RoKu1) [reply src/lib.rs:264]: Yep seems fine, can I do it as part of other PR? or this one itself.

Comment (sticnarf) [reply src/lib.rs:264]: It's okay to just to do it in this PR.

Comment (RoKu1) [reply src/lib.rs]: Done.

File: src/tcp/mod.rs [hunk 2]
```diff
@@ -0,0 +1,2 @@
+pub mod socks4;
+pub mod socks5;
```
Comment (sticnarf) [inline src/tcp/mod.rs:2]: To keep compatible with old versions, please re-export the inner types to `tcp` here.

File: src/tcp/socks4.rs [hunk 3]
```diff
@@ -0,0 +1,519 @@
...
+    /// `TargetAddr`.
+    pub async fn connect<'t, P, T>(proxy: P, target: T) -> Result<Socks4Stream<TcpStream>>
+    where
+        P: ToProxyAddrs,
+        T: IntoTargetAddr<'t>,
+    {
+        Self::execute_command(proxy, target, None, CommandV4::Connect).await
+    }
+
+    /// Connects to a target server through a SOCKS5 proxy using given username,
```
Comment (sticnarf) [inline src/tcp/socks4.rs:62]:
```suggestion
    /// Connects to a target server through a SOCKS4 proxy using given username,
```

Comment (sticnarf) [reply src/tcp/socks4.rs:62]: Same for a few other similar cases in the file

File: src/tcp/mod.rs (hunk 2 above)
Comment (RoKu1) [reply src/tcp/mod.rs:2]: Sure done

File: src/tcp/socks4.rs (hunk 3 above)
Comment (RoKu1) [reply src/tcp/socks4.rs]: Done

**Review by sticnarf — APPROVED** [review]
Great job! Thanks.

**Pull request merged**